updatehub/tests/common.rs

```diff
@@ -0,0 +1,273 @@
+// Copyright (C) 2020 O.S. Systems Sofware LTDA
+//
+// SPDX-License-Identifier: Apache-2.0
+
+use mockito::{mock, Mock};
+use regex::Regex;
+use serde_json::json;
+use std::{env, path::PathBuf};
+
+pub enum FakeServer {
+    NoUpdate,
+    HasUpdate(String),
+}
+
+pub enum Polling {
+    Enable,
+    Disable,
+}
+
+pub enum Server {
+    Custom(String),
+    Standard,
+}
+
+pub struct Settings {
+    pub polling: bool,
+    pub listen_socket: String,
+    pub server_address: String,
+    pub download_dir: Option<PathBuf>,
+    pub config_file: Option<PathBuf>,
+}
+
+impl Default for Settings {
+    fn default() -> Self {
+        Settings {
+            polling: false,
+            listen_socket: String::from("localhost:8080"),
+            server_address: mockito::server_url(),
+            download_dir: None,
+            config_file: None,
+        }
+    }
+}
+
+impl Settings {
+    pub fn init_server(self) -> (rexpect::session::PtySession, updatehub::tests::TestEnvironment) {
+        let setup = updatehub::tests::TestEnvironment::build()
+            .listen_socket(self.listen_socket)
+            .server_address(self.server_address)
+            .add_echo_binary("reboot");
+
+        let mut setup = if !self.polling { setup.disable_polling() } else { setup }.finish();
+
+        if let Some(download_dir) = self.download_dir {
+            setup.settings.data.update.download_dir = download_dir;
+
+            let content =
+                toml::ser::to_string_pretty(&setup.settings.data.update.download_dir).unwrap();
+            std::fs::write(&setup.settings.stored_path, content).unwrap();
+        }
+
+        if let Some(config_file) = self.config_file {
+            setup.settings.stored_path = config_file;
+        }
+
+        println!("{:#?}", setup.settings.data);
+
+        let cmd = format!(
+            "{} server -v trace -c {}",
+            cargo_bin("updatehub").to_string_lossy(),
+            setup.settings.stored_path.to_string_lossy()
+        );
+
+        let handle = rexpect::spawn(&cmd, None).expect("fail to spawn server command");
+
+        (handle, setup)
+    }
+
+    pub fn config_file(self, p: PathBuf) -> Self {
+        Settings { config_file: Some(p), ..self }
+    }
+
+    pub fn download_dir(self) -> Self {
+        let tmp_dir = tempfile::tempdir().unwrap();
+        let mut perms = std::fs::metadata(tmp_dir.path()).unwrap().permissions();
+        perms.set_readonly(true);
+        std::fs::set_permissions(tmp_dir.path(), perms).unwrap();
+
+        let tmp_dir = Box::leak(Box::new(tmp_dir));
+
+        Settings { download_dir: Some(tmp_dir.path().to_owned()), ..self }
+    }
+
+    pub fn polling(mut self) -> Self {
+        self.polling = true;
+        self
+    }
+
+    pub fn listen_socket(self, s: String) -> Self {
+        Settings { listen_socket: s, ..self }
+    }
+
+    pub fn server_address(self, s: String) -> Self {
+        Settings { server_address: s, ..self }
+    }
```

Comment:
I think we should use the same mutation patter on all methods. I'm fine with either but keeping a sigle pattern is important.